### DISCUSSION TITLE
Help with APA107/APA102 Setup

### DISCUSSION THREAD

**Discussion by openupshop** [awawa-dev/HyperHDR#212] [Q&A] [answered]
Hello All.  Please bare with me as I am somewhat new to this.  So after being disappointed with the $150 GOVEE immersion kit from amazon, I very happily came across HYPERHDR.  I

My setup:   APA107 LED's, [LINK](https://www.amazon.com/APA107-Smart-Pixel-Addressable-Waterproof/dp/B07MDD191C/ref=sr_1_2?crid=1AZLV8K74RSAL&keywords=apa107&qid=1645047497&sprefix=apa107%2Caps%2C122&sr=8-2&th=1).  
Raspberry pi 4 8gb ram flashed with the 64 bit HyperHDR SD Card image:  [SD-card-image-rpi234-aarch64.zip](https://github.com/awawa-dev/HyperHDR/releases/download/v17.0.0.0/SD-card-image-rpi234-aarch64.zip) and I have updated to the latest version of HYPERHDR.
4KVC00 capture card [Link](https://www.amazon.com/AV-Access-Nintendo-Streaming-Recording/dp/B08MPPKZ55/ref=sr_1_1?crid=S1HFT7X0NWVS&keywords=B08mppkz55&qid=1645046797&sprefix=b08mppkz55%2Caps%2C108&sr=8-1) 
5V 40A power supply 
Source:  Laptop for testing, AVR with dual HDMI outputs as my eventual source once I get through testing.  Here is my setup, please excuse the childish drawing:

![Screenshot_20220216-142907_Gallery#1](https://user-images.githubusercontent.com/99840700/154370706-ff023a9b-d3e4-4422-9e7d-18e4f8050a28.jpg)

I am not using my AVR as the source yet, as I am just testing things out, hopefully I can avoid having to get a splitter if the 2nd HDMI output of my AVR works well.  For now, I am using my laptop as source going to the capture card.  Anyhow, I was hoping to get some help/recommendations with a couple of things.  

1) When I run a youtube ambilight video, it works great so far, the leds respond instantly to color changes although I have yet to test HDR.  However, when I shutdown the computer or disconnect the source, the led's go out, but come back on after a bit with a frozen rainbow, then go out again, and on and on.

2) I see a file [V17_LUT_table_for_4KVC00_1080pHDR.tar.xz](https://github.com/awawa-dev/HyperHDR/releases/download/v17.0.0.0/V17_LUT_table_for_4KVC00_1080pHDR.tar.xz) in the HYPERHDR release page which is for my capture card.  What exactly do I do with this or how do I install it or apply it??

3) Lastly the LEDS are listed as "APA107's on the amazon page and on box it came in, but on the actual roll, it says APA102.  Any particular settings I should change in HYPERHDR, as there was not APA107 listed on the controller section, I chose APA102 anyhow.

Thank you very much guys/gals.  Truly a pleasure to hopefully be able to pick some of your brains about this....

**Comment by awawa-dev** [comment, marked as answer]
Hi
1 that grabber is a bit problematic, it simply disconnects when it looses video source. Or even when the video source resolution or refresh rate is changing...usually it backs after few seconds and 'auto-resume' could help for that short period of absence. For frozen rainbow have no idea. Verify HyperHDR logs and ```dmesg```. That grabber leaves a lot of trace.
2 Yes, absolutely for that grabber for HDR tone mapping.  Extract it and put it in the hyperhdr home directory (.hyperhdr) the path depends on the OS, but that path is written in first lines in logs. Next time you activatye HDR tone mapping in logs you will receive confirmation if it was found or not.
3 Chinese manufactures are very creative, APA107 seems similar to much better HD107 ;) I suspect it is not true APA102 (very, very hard to find genuine one)  but clone sk9822. But it should work with APA102 or SK9822 (test that one) driver. 
4 Powering Rpi4 using LED strip power supply is NOT recommended. Sometimes it is causing very hard to diagnose problems with the hardware (most often with USB grabber). Everybody wants to simplify their setups but definitely that's not proper&safe way.

**Reply by openupshop** [reply]
Hello. Thank you so much for the detailed reply, truly appreciate it.🙏🙏 I can def post logs, but before trying that since you mentioned the "grabber being problematic," would you recommend I just buy a diff grabber. I got it from Amazon for ~$50 & I'm well within the return period so getting a diff one wouldn't be much of an issue.  Also I did come across this bug report [Link](https://github.com/hyperion-project/hyperion.ng/issues/1007) as far as the "frozen rainbow" & they mentioned having fixed it on the next update, so not sure why I am still getting it.

And thank you, your instructions for the file are crystal clear.

I'm sorry my childish drawing had one thing incorrect, I am actually powering the pi with a separate 5V 3Amp power supply.

Lastly, I did read a ton about fake APA102's prior to this & assumed this roll was probably some Chinese clone, a little sad because I had paid $67 when i purchased the strip. And once I began this project, I even purchased a roll of WS2815B's as a backup, [Link](https://www.amazon.com/your-orders/pop/ref=ppx_yo_mob_b_pop?_encoding=UTF8&gen=canonical&lineItemId=nhsoroirrpqnsny&orderId=111-9090087-9585051&packageId=1&returnSummaryId=XryFa7E04n6dmjU%2BRHTDcA&returnUnitIndices=1&returnUnitMappingId=1011987972899%231&shipmentId=Txbm95yn4)
But once I saw these "APA102's" lit, their color & brightness were very impressive compared to most led strips I've had for various projects.  Anyhow, I say all that to ask if there another strip you recommend as far as what would be your number one choice?  I have read that the [SK6812's](https://www.amazon.com/BTF-LIGHTING-Individually-Addressable-Flexible-Waterproof/dp/B01MTW2L6B?th=1) are a good choice since in addition to RGB, it has the the white led.  

Thank you again, really appreciate your help & feedback.🙏🙏🙏

**Reply by awawa-dev** [reply]
Don't get me wrong: that grabber provides very good image quality especially for HDR with special LUT table and in certain conditions it works very well. But...provided USB cable is not good, it's too long often causing errors in the transmission so it is recommended to replace it with a solid USB3.1 gen 2 cable. Connected to USB2.0 (not your case) it's unstable due to high power consumption. We can fix connection issue for resolution/refreshrate changing with auto-resume option, but if the HDMI source is not available when HyperHDR starting the application could simply not detect it. And under Linux it can only capture using 1080p60 (1080p30 doesn't work) which is causing high CPU load. So there is a lot of conditions that must be taken into account. 

LED drivers handling were completely rebuilt, especially turn on/off mechanism for all of them and will be provided as v18beta2 so I hope it will fix APA102 issue if it still persists.

You find a method how to verify if your APA102 is genuine here: https://www.hyperhdr.eu/2020/11/blog-post.html take a close look at your LED and check that microchip inside it. Compare it to the picture on the blog.

I'm great fan of RGBW (cold white version) LED strips. Currently only popular SK6812 is available on the market in the acceptable prices. But you can't drive SK6812 directly from Rpi4 (ok, you can and HyperHDR has a proper driver but it's definitely not recommended or supported by me) and external driver like proper ESP32 or ESP8266 board is needed. Read more on HyperSerialEsp8266/HyperSerialEsp32/HyperserialWled or HyperSpi. You also need a level shifter (even for APA102 driven by Rpi4) but sometimes it works without it but be aware of possible problems for example: sudden burst of light due to instability.

**Reply by openupshop** [reply]
Can't say thank you enough, you were so informative & helpful.  After following your recommendation for a short USB3.1 gen 2 cable, the lighting has worked almost flawlessly including (for now) staying off when there's no signal.  🙏🙏

You mentioned the capture card in linux being limited to1080p 60 fps causing high CPU load.  Do you think this is a bit mitigated since I am using a PI 4 with 8GB ram?  

I had seen the blog with the microchip pics prior to starting, and why I got the WS2815B's.  Mine are def not the original/genuine APA102's.  But whatever they are, they've been great so far.   But in case longevity becomes an issue, is there a particular led type you recommend since you said you don't recommend the SK6812's??

Thank you again, truly appreciate it🙏🙏

**Reply by awawa-dev** [reply]
> You mentioned the capture card in linux being limited to1080p 60 fps causing high CPU load. Do you think this is a bit mitigated since I am using a PI 4 with 8GB ram?

HyperHDR on Rpi4 can handle that stream (quarter of frame mode option should be enabled anyway) but using such high resolution & high frame rate stream is completely unnecessary for our ambient lightning: causes high CPU usage and introduces some lag. And since that grabber offers 1080p30 in the available mode list but that mode doesn't work then it's a bug in the Linux driver or in the USB grabber firmware. 
It is a secondary question how much memory is available to us if only one application (HyperHDR) is running on the RPI: even a 512MB board is enough for a HyperHDR running on the Raspberry Pi OS CLI (without X desktop).

>  But in case longevity becomes an issue, is there a particular led type you recommend since you said you don't recommend the SK6812's??

Maybe I didn't make myself clear, but I definitely recommend RGBW SK6812 cold version for the visual effect. This is also stated in the wiki introduction: https://github.com/awawa-dev/HyperHDR/wiki Longevity cannot be guaranteed, I had one setup where the SK6812 died fairly quickly but others have been around for over a year now (heavy use).

**Comment by openupshop** [comment]
Yea that makes perfect sense since obviously the ambient lighting simply needs the color info, even 720p seems like it would be more than enough. 

And sorry I actually misunderstood your statement of _"(ok, you can and HyperHDR has a proper driver but it's definitely not recommended or supported by me)"_ as you saying you do not recommend the SK6812's at all when now I understand you meant you don't recommend "directly" driving them from the pi.

Although the 2nd HDMI output of my AVR to the grabber worked great...after a ton of research & testing, I realized once my AVR is outputting two HDMI signals at once, HDR is automatically disabled.  As an alternative, I am now using the HDMI passthrough of that grabber to my tv, and the leds, responsiveness, color, all are working better than I could have ever expected... However, after about a few minutes, my tv loses the hdmi signal while the ambient lighting continues to work great.  I ordered much shorter 40 GBPS certified HDMI cables to see if that helps.  I also ordered a very short USB 3.1 gen 2 cable per your recommendations to get rid of that long troublesome cable that came with the grabber (also per your recommendation, I ordered the level shifter).  Hopefully, the new hdmi cables fix the signal loss issue. 

Thank you again my friend, you were not only really helpful, I learned a ton and I am certain this thread will help others in the future who have similar issues.  Thank you 🙏🙏🙏🙏

**Reply by awawa-dev** [reply]
> And sorry I actually misunderstood your statement of "(ok, you can and HyperHDR has a proper driver but it's definitely not recommended or supported by me)" as you saying you do not recommend the SK6812's at all when now I understand you meant you don't recommend "directly" driving them from the pi.

Exactly. You should drive SK6812 or WS2812b using external ESP32 or ESP8266.   
SK6812/WS2812b protocol is much more "complicated" than SPI driven APA102 and the device must to emulate it & keep proper timings. ws2812 library on Raspberry Pi takes some resources and is rather problematic in use. It is much better to delegate that task to specialized external unit (ESP) that only takes care about one thing.

**Reply by openupshop** [reply]
Thank you.  Just for future ambient enthusiasts & clarification, is the protocol for WS2815B complicated or are those fine to run of the Pi directly?  Based on your explanations, I would only switch to SK6812, they seem like the best option but I had ordered the WS2815B's as a backup to my current setup so I am asking in case someone else sees this thread.  Also some may want the WS2815B if they have a long run due to it being 12 Volts so hence much less voltage drop vs 5 volts led's.    Thank you again, your posts are so informative.

**Reply by awawa-dev** [reply]
Most sk6812 setups are fine with one input, to be safe it's recommended to put 2 power injection: at the start and at the end. You can provide as much power injections as you want. WS2815B  protocol is just same story as ws2812/sk6812: there is no separate clock line (only second backup data line) and everything hangs in keeping with strictly defined timings.

**Reply by openupshop** [reply]
Thank you, that is not only great info for me, however also for anyone coming across this page in the future.  

Additionally, for anyone else.... concerning my issue of losing tv signal after several minutes using the HDMI passthrough of the 4KVC00 grabber - after a ton of searching, I realized why this is happening.  So despite having 4k @60, passthrough, and all the other features, the one thing the 4KVC00 lacks, is HDCP compliance. I had HDCP compliance as a must have spec along with 4K passthrough, etc etc. in my head when I was searching countless grabbers, guess I just missed it.  Also, kind of stupid it has all the other features but no HDCP.  Anyhow, the fact that it works for several minutes prevented my brain from realizing HDCP compliance being the issue, I had always (very incorrectly) assumed if a device lacked HDCP, it just wouldn't work.   Of course, unless I am wrong about HDCP being the issue, but it is missing on their spec page which I will attach.

![pasted-from-clipboard](https://user-images.githubusercontent.com/99840700/156242851-e9e74cc9-165d-432f-bec5-f5a2e51b6598.png)

As a "hopeful" solution, I have order [Link to Feintech VSP01222](https://www.amazon.co.uk/FeinTech-VSP01222-Splitter-Outputs-Receiver/dp/B08KFD4XGK).  Had to order it from Amazon UK, came out to roughly $80 USD.. can't believe the Feintech items are not yet available in the US or directly from their website to US.  Overall it seems to have all the major features, 4K @ 60, HDCP 2.2, Scaler, etc, but most important of all where virtually all the Non-$400/$500 splitters lack ARC, CEC or both, this has it.  From Specs listed:  ARC pass
                               _If you are using cable TV, satellite TV or smart TV apps on your TV, the sound can be transmitted 
                               to the sound system via HDMI ARC as usual. The control commands of the TV remote control (HDMI-CEC)
                               are also allowed to pass through to the sound system._

Also for anyone who has an Audio/Video Receiver & wants to setup ambient lighting with HYPERHDR, the setup picture with this grabber that is attached will be a treat for your eyes as getting both ARC & CEC to work with an ambient system has been very difficult outside of spending $400/$500 for a splitter.  Hopefully, the VSP01222 doesn't come with any major issues of its own, but I will post back once I receive it & set thing up.

Thank you again AWAWA, of course none of this would be possible without you....

<img width="692" alt="Screen Shot 2022-03-01 at 12 27 06 PM" src="https://user-images.githubusercontent.com/99840700/156244661-216aca66-6ea4-4ee9-a36c-2ce873050d43.png">
